---
title: "Data Analysis for Early and Late Lung-colonizing Tumors"
author: "Emily Franz and Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code_folding: hide
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    echo = TRUE,
    cache = TRUE,
    collapse = TRUE,
    tidy.opts = list(width.cutoff = 95),
    message = FALSE,
    warning = FALSE,
    cache.lazy = FALSE
)
```

```{r lib}
# Load packages
library(Seurat)
library(tidyverse)
library(rrrSingleCellUtils)
theme_set(theme_bw())
theme_update(plot.title = element_text(hjust = 0.5))
library(patchwork)
library(msigdbr)

# Set random generator seed to facilitate reproducibility
set.seed(888)
```

## Make up directory structure
```{bash mkdirs, eval=TRUE}
for directoryName in \
  output \
  output/figures \
  output/rdata \
  output/de
do
  if [ ! -d ${directoryName} ]
  then
    mkdir ${directoryName}
  fi
done
```

# Load and Process Datasets
Datasets are from tumors collected from mouse lungs at early (day 14 post injection of tumor cells) and late (day 35 post injection of tumor cells) timepoints. mCherry was used to label the tumor cells as well as the surrounding niche cells.
```{r load}
sample_list <- read_tsv("misc/sample_cutoffs.txt", show_col_types = FALSE)

sobj_list <- list()

for (i in seq_len(nrow(sample_list))) {
    species_pattern <- sample_list$species_pattern[i]
    obj_name <- sample_list$obj_name[i]

    sobj_list[[obj_name]] <-
        tenx_load_qc(paste0("/home/gdrobertslab/lab/Counts/",
                            sample_list$sample_name[i],
                            "/filtered_feature_bc_matrix"),
                     violin_plot = FALSE,
                     species_pattern = species_pattern)

    # Add metadata to dataset
    for (colname in colnames(sample_list)) {
        sobj_list[[obj_name]][[colname]] <- sample_list[[colname]][i]
    }

    cutoff_tibble <-
        tribble(~feature, ~min_val, ~max_val,
                "nCount_RNA",
                sample_list$ncount_rna_min[i],
                sample_list$ncount_rna_max[i],

                "percent.mt",
                0,
                sample_list$mt_max[i])

    plotted <-
        feature_hist(sobj_list[[obj_name]],
                     features = c("nCount_RNA", "percent.mt"),
                     cutoff_table = cutoff_tibble)

    ggsave(paste0("output/figures/feature_hist_",
                  obj_name,
                  ".png"),
           width = 10,
           height = 10,
           plot = plotted)

    sobj_list[[obj_name]] <-
        sobj_list[[obj_name]] %>%
        subset(nCount_RNA >= sample_list$ncount_rna_min[i] &
               nCount_RNA <= sample_list$ncount_rna_max[i] &
               percent.mt <= sample_list$mt_max[i]) %>%
        process_seurat()
}
save(sobj_list,
     file = "output/rdata/sobj_list.rds")

```

## UMAP with all samples
```{r umap_alldependson='load', cache.vars=''}
load("output/rdata/sobj_list.rds")
all_samples <-
    merge(sobj_list[[1]],
          y = sobj_list[2:length(sobj_list)],
          add.cell.ids = names(sobj_list)) %>%
    process_seurat(resolution = 0.2)

plot_name <-
    DimPlot(all_samples,
            group.by = "sample_description",
            split.by = "sample_description",
            shuffle = TRUE,
            ncol = 2,
            cols = plot_cols[seq_along(sobj_list)],
            pt.size = 0.01,
            order = FALSE) +
    theme_roberts() +
    labs(title = NULL)

ggsave("output/figures/umap_all_samples.pdf",
       width = 4,
       height = 6,
       plot = plot_name)


```

## Human Cell Only Processing
```{r human-processing, dependson="load"}
# Human
# Normalize the number of cells per sample
n_cells <- c()

for (i in seq_along(sobj_list)) {
    n_cells[i] <- length(colnames(sobj_list[[i]]))
}

names(n_cells) <- names(sobj_list)

# Merge day 14 and day 28 Seurat objects together
tumor <-
    merge(sobj_list[["d14_pos_h"]],
          y = sobj_list[c("d28_pos_h", "tibia_tumor_h")],
          add.cell.ids = c("d14", "d28", "tibia"),
          project = "Early vs Late OS Mets") %>%
    NormalizeData() %>%
    ScaleData() %>%
    FindVariableFeatures() %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    FindClusters(resolution = 0.15, k.param = 30) %>%
    RunUMAP(dims = 1:30, n.neighbors = 30L, metric = "euclidean")

# Save human mCherry+ merged object
save(tumor, file = "output/rdata/d14_d28_tumor.RData")

# Use timepoint naming (Day 14/Day 28) for UMAP plotting
umap_timepoint <-
    r_dim_plot(tumor,
        title = "Tumor at metastatic timepoints",
        group.by = "timepoint",
        label = TRUE)
umap_timepoint

pdf("output/figures/d14_d28_umap.pdf",
    width = 5,
    height = 5)
umap_timepoint
dev.off()

# Use basic naming (early/late) for UMAP plotting
umap_group <-
    r_dim_plot(tumor,
        title = "Tumor at metastatic timepoints",
        group.by = "group",
        label = TRUE,
        repel = TRUE)
umap_group

pdf("output/figures/d14_d28_umap.pdf",
    width = 5,
    height = 5)
umap_group
dev.off()

# Reorder the group metadata variable
tumor$group <- factor(tumor$group, levels = c("Primary", "Early", "Late"))

umap_group_split <-
    r_dim_plot(tumor,
        title = "Tumor at metastatic timepoints",
        split.by = "group",
        label = TRUE,
        repel = TRUE)
umap_group_split

```

## Run DE on human cells for GSEA and IPA
```{r human-de, dependson="load"}
load("output/rdata/sobj_list.rds")

tumor_three_tp <-
    merge(sobj_list[["d14_pos_h"]],
          y = sobj_list[c("d28_pos_h", "d49_pos_h", "tibia_tumor_h")],
          add.cell.ids = c("d14", "d28", "d49", "tibia")) %>%
    process_seurat(resolution = 0.15, run_umap_dims = 1:20)

# Save human mCherry+ merged object
qs::qsave(tumor, file = "output/rdata/d14_28_49_tumor.qs")
```

## Mouse Cell Only Processing
Removing this chunk as it's not used anywhere
```{r mouse-processing, dependson="load"}
# Mouse
# Merge day 14 and day 35 Seurat objects together
# microenv <-
#     merge(sobj_list[["d14_pos_m"]],
#           y = sobj_list[c("d28_pos_m", "d49_pos_m", "S0006_mouse")],
#           add.cell.ids = c("d14", "d28", "d49", "S0006"),
#           project = "Early vs Late OS Mets")

# # Process merged object
# microenv <-
#     NormalizeData(microenv) %>%
#     ScaleData() %>%
#     FindVariableFeatures() %>%
#     RunPCA() %>%
#     FindNeighbors(dims = 1:30) %>%
#     FindClusters(resolution = 0.15, k.param = 30) %>%
#     RunUMAP(dims = 1:30, n.neighbors = 30L, metric = "euclidean")

# # Plot UMAP
# env_umap <- r_dim_plot(microenv,
#         title = "Stroma at metastatic timepoints",
#         label = TRUE)
# env_umap

# env_umap_split <- r_dim_plot(microenv,
#         title = "Stroma at metastatic timepoints",
#         split.by = "group")
# env_umap_split

# timepoints_panel <- (umap_group | umap_group_split) / (env_umap | env_umap_split)
# timepoints_panel

# # Save mouse mCherry+ merged Seurat object
# save(microenv, file = "output/rdata/early_late_microenv.RData")
```

# Plotting Expression: Genes of Interest

## Human Cells Only: Pre-cell cycle regression
```{r human_geneplots, dependson=c("load","human-processing")}
# Dot plot of expression of genes of interest by group (early/late)
gene_list <-
    c("BCL2",
      "BCL2L1",
      "BCL2L2",
      "MCL1")

dot_plot <-
    DotPlot(tumor,
            features = gene_list,
            group.by = "group",
            scale = FALSE,
            cols = c("blue", "red")) +
  RotatedAxis() +
  coord_fixed()

# Save as pdf
pdf("output/figures/early_late_dot_plot.pdf",
    width = 7,
    height = 4)
dot_plot
dev.off()

# Violin plot of expression of genes of interest by group (early/late)
vln_plot <-
    VlnPlot(tumor,
            features = gene_list,
            group.by = "group") +
    RotatedAxis()

# Save as pdf
pdf("output/figures/early_late_vln_plot.pdf",
    width = 6,
    height = 5)
vln_plot
dev.off()

# Feature plot (UMAP) of expression of genes of interest by group (early/late)
feature_plot <-
    FeaturePlot(tumor,
                features = gene_list,
                split.by = "group",
                order = TRUE)

# Save as pdf
pdf("output/figures/early_late_feature_plot.pdf",
    width = 6,
    height = 12)
feature_plot
dev.off()
```

## Human Cells Only: Post-cell cycle regression

Regress out cell cycle genes so that the dataset does not cluster on the UMAP 
by cell cycle gene expression.

```{r cc, dependson=c("load","human-processing")}
# Calculate cell cycle phase and view its effects on clustering
# Note: default for kill_cc() is not performing cell cycle regression

# Save as pdf
pdf("output/figures/early_late_tumor_pre-kill_cc.pdf",
    width = 5,
    height = 5)
kill_cc(tumor)
dev.off()

# Calculate cell cycle phase and remove its effects on clustering
tumor_cc <-
    kill_cc(tumor,
            cc_regress = "Y") %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.15) %>%
    RunUMAP(dims = 1:20)

# Plot tumor UMAP after cell cycle genes are regressed out
DimPlot(tumor_cc,
        group.by = "Phase",
        label = TRUE,
        pt.size = 1) +
    theme_roberts()

# Save as pdf
pdf("output/figures/early_late_post-kill_cc.pdf",
    width = 5,
    height = 5)
DimPlot(tumor_cc,
        group.by = "Phase",
        label = TRUE,
        pt.size = 1) +
    theme_roberts()
dev.off()

# Plot tumor UMAP by timepoint post cell cycle regression
umap_cc <-
    DimPlot(tumor_cc,
            pt.size = 1,
            label = TRUE,
            group.by = "timepoint") +
    theme_roberts() +
    theme(aspect.ratio = 1)
umap_cc

# Save as pdf
pdf("output/figures/early_late_umap_cc_bytrtment.pdf",
    width = 5,
    height = 5)
umap_cc
dev.off()

# Save the cell cycle regressed tumor object
save(tumor_cc, file = "output/rdata/early_late_tumor_cc.RData")
```

Repeat plots of gene expression post cell cycle regression.
Add S0131 to dotplot
```{r human_geneplots_postcc, dependson=c("load","human-processing","cc")}
# Dot plot of expression of genes of interest by group (early/late)
dot_plot <-
    DotPlot(tumor_cc,
            features = gene_list,
            group.by = "group",
            scale = FALSE,
            cols = c("blue", "red")) +
  RotatedAxis() +
  coord_fixed()

# Save as pdf
pdf("output/figures/early_late_dot_plot_postccr.pdf",
    width = 7,
    height = 4)
dot_plot
dev.off()

# Violin plot of expression of genes of interest by group (early/late)
vln_plot <-
    VlnPlot(tumor_cc,
            features = gene_list,
            group.by = "group") +
  RotatedAxis()

# Save as pdf
pdf("output/figures/early_late_vln_plot_postccr.pdf",
    width = 6,
    height = 5)
vln_plot
dev.off()

# Feature plot (UMAP) of expression of genes of interest by group (early/late)
feature_plot <-
    FeaturePlot(tumor_cc,
                features = gene_list,
                split.by = "group",
                order = TRUE) +
  theme(aspect.ratio = 1)

# Save as pdf
pdf("output/figures/early_late_feature_plot_postccr.pdf",
    width = 6,
    height = 12)
feature_plot
dev.off()
```

## Growth factor expression in normal vs lungs with mets
Also plot each time point separately as a time-course
```{r deGrowthFactorMets, dependson='load'}
cm_target_genes <-
    read_tsv("misc/growth_factor_genes.txt",
             col_names = "human_gene",
             show_col_types = FALSE) %>%
    left_join(read_tsv("misc/HMD_HumanPhenotype.rpt",
                       col_names = c("human_gene",
                                     "h_entrez",
                                     "mouse_gene",
                                     "mgi",
                                     "pheno",
                                     "junk"),
                       show_col_types = FALSE) %>%
        select(human_gene, mouse_gene),
        by = "human_gene") %>%
    select(mouse_gene)

load("output/rdata/sobj_list.rds")

samples <- c("S0006_mouse", "d14_pos_m", "d28_pos_m", "d49_pos_m")

sobj_list$S0006_mouse$group_name <- "Control lung"

for (sample_name in c("d14_pos_m", "d28_pos_m", "d49_pos_m")) {
    sobj_list[[sample_name]]$group_name <- "Met. lung"
}

combined_data <-
    merge(x = sobj_list$S0006_mouse,
          y = list(sobj_list[["d14_pos_m"]],
                   sobj_list[["d28_pos_m"]],
                   sobj_list[["d49_pos_m"]])) %>%
    process_seurat()

combined_data$sample_description <-
    factor(combined_data$sample_description,
           levels = c("SCID control lung",
                      "OS17 mets - 2wk",
                      "OS17 mets - 4wk",
                      "OS17 mets - 7wk"))

combined_data$sample_name <-
    combined_data$sample_name %>%
    str_remove("/outs") %>%
    factor(., levels = c("S0006-WTLung",
                         "S0082",
                         "S0084",
                         "S0093"))
```

### Run DE on the mouse samples
```{r deGrowthFactorMetsDE, dependson='deGrowthFactorMets'}
Idents(combined_data) <- combined_data$group_name

overall_de <-
    FindMarkers(combined_data,
                ident.1 = "Control lung",
                logfc.threshold = 0)
```

### Get reference datasets
GSE151974
```{r mouse_ref_GSE151974}
ref_path <- "/home/gdrobertslab/lab/GenRef/sc_ref_datasets/mouse"

data_matrix <-
    read.csv(paste0(ref_path,
             "/GSE151974/GSE151974_raw_umi_matrix_postfilter.csv.gz"),
             header = TRUE) %>%
    column_to_rownames("X")

metadata <-
    read.csv(paste0(ref_path,
             "/GSE151974/GSE151974_cell_metadata_postfilter.csv.gz"),
             header = TRUE) %>%
    column_to_rownames("X")

mouse_lung_ref <-
    CreateSeuratObject(counts = data_matrix,
                       project = "mouse_lung_ref",
                       meta.data = metadata) %>%
    subset(Oxygen == "Normoxia") %>%
    process_seurat(resolution = 0.15, run_umap_dims = 1:20)
mouse_lung_ref$cell_type <- mouse_lung_ref$CellType
mouse_lung_ref$dataset <- "GSE151974"

DimPlot(mouse_lung_ref,
        pt.size = 1,
        label = TRUE,
        group.by = "CellType") +
    NoLegend()

qs::qsave(mouse_lung_ref, file = "output/rdata/mouse_lung_ref.qs")
```

### Add reference labels to mouse cells
```{r mouse_ref_labels}
mouse_lung_ref <- qs::qread("output/rdata/mouse_lung_ref.qs")

cancer_imm <- readRDS("/home/gdrobertslab/lab/GenRef/sc_ref_datasets/mouse/tumor_imm_atlas/TICAtlas_downsampled.rds")
cancer_imm_data <-
    GetAssayData(cancer_imm) %>%
    as.matrix() %>%
    as.data.frame() %>%
    mutate(gene = rownames(cancer_imm) %>%
            nichenetr::convert_human_to_mouse_symbols() %>%
            as.vector()) %>%
    filter(!is.na(gene)) %>%
    group_by(gene) %>%
    slice_head(n = 1) %>%
    column_to_rownames("gene") %>%
    as.matrix()

mouse_immune <- celldex::ImmGenData()
# This label is not informative
mouse_immune <-
    mouse_immune[mouse_immune$label.main != "Stromal cells", ]

mouse_rna <- celldex::MouseRNAseqData()

cell_assign <-
    SingleR::SingleR(as.SingleCellExperiment(combined_data),
                     ref = list(GetAssayData(mouse_lung_ref),
                                cancer_imm_data,
                                mouse_immune,
                                mouse_rna),
                     labels = list(mouse_lung_ref$cell_type,
                                   cancer_imm$lv2_annot,
                                   mouse_immune$label.main,
                                   mouse_rna$label.fine),
                     aggr.ref = TRUE,
                     num.threads = 1)

SingleR::plotScoreHeatmap(cell_assign)

combined_data$cell_type <- cell_assign$labels

combined_data$cell_score <- cell_assign$scores %>%
    apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

set.seed(1337)
celltype_mouse_plot <-
    DimPlot(combined_data,
            group.by = "cell_type",
            split.by = "group",
            label = TRUE,
            repel = TRUE) +#,
    #        cols = sample(rainbow(n = length(unique(combined_data$cell_type))))) +
        NoLegend()

celltype_mouse_plot_all <-
    DimPlot(combined_data,
            group.by = "cell_type",
            label = TRUE,
            repel = TRUE) +#,
    #        cols = sample(rainbow(n = length(unique(combined_data$cell_type))))) +
        NoLegend()

two_plots <-
    wrap_plots(celltype_mouse_plot_all,
               celltype_mouse_plot,
               ncol = 2,
               widths = c(1, 4))


ggsave("output/figures/mouse_umaps_celltype_2.pdf",
       width = 20,
       height = 8,
       plot = two_plots)

test <- process_seurat(combined_data, resolution = 1)

table(test$cell_type, test$seurat_clusters) %>%
    as.data.frame() %>%
    dplyr::rename(cell_type = Var1,
                  cluster = Var2,
                  count = Freq) %>%
    ggplot(aes(x = cluster,
               y = cell_type,
               fill = log10(count),
               label = count)) +
    geom_tile() +
    geom_text()

DimPlot(test, cols = sample(rainbow(n = length(unique(test$cell_type)))))
DimPlot(test,
        cells.highlight = WhichCells(test,
                                     expression = seurat_clusters == 16))

combined_data$group <-
    factor(combined_data$group,
           levels = c("Mouse_control_lung",
                      "Early_mouse_stroma",
                      "Midpoint_mouse_stroma",
                      "Late_mouse_stroma"))


harmonization <-
    tribble(~from,      ~to,
            "Alv Mf",               "Alveolar macrophages",
            "Int Mf",               "Interstital macrophages",
            "DC 1",                 "Dendritic cells",
            "DC 2",                 "Dendritic cells",
            "Art",                  "Arterial endothelial cells",
            "Vein",                 "Venous endothelial cells",
            "Lymph",                "Lymphatic endothelial cells",
            "Mono",                 "Monocytes",
            "NK cell",              "Natural killer cells",
            "natural killer cells", "Natural killer cells",
            "ILC2",                 "Innate lymphoid cells",)

```

### Count cell types
```{r deGrowthFactorMetsCellType, dependson='deGrowthFactorMetsDE'}
# imm_cells <- celldex::ImmGenData()
# mouse_rna <- celldex::MouseRNAseqData()

# cell_assign <-
#     SingleR::SingleR(as.SingleCellExperiment(combined_data),
#                      ref = list(imm_cells,
#                                 mouse_rna),
#                      labels = list(imm_cells$label.main,
#                                    mouse_rna$label.main))

# combined_data$cell_type <-
#     cell_assign$labels %>%
#     str_replace("DC", "Dendritic cells")
# combined_data$cell_score <-
#     cell_assign$scores %>%
#     apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

cell_type_list <-
    table(combined_data$cell_type, combined_data$group) %>%
    as.data.frame() %>%
    group_by(Var1) %>%
    summarize(sum = sum(Freq), .groups = "drop") %>%
    filter(sum > 50)
```

### Make supplemental figures with all cell types
```{r deGrowthFactorMetsLoop, dependson='deGrowthFactorMetsCellType'}
target_genes <- cm_target_genes

target_genes <-
    left_join(target_genes,
              overall_de %>%
              rownames_to_column("mouse_gene"))

################################################################################
# DE between lung and primary mets by cell type
combined_subset <-
    combined_data %>%
    subset(cell_type %in% cell_type_list$Var1)

combined_subset <-
    combined_subset[rownames(combined_subset) %in% target_genes$mouse_gene, ]

Idents(combined_subset) <- combined_subset$group_name

qs::qsave(combined_subset,
          file = "output/rdata/combined_subset.qs")

cell_type_de <- tibble(gene = character())

for (cell_type in unique(combined_subset$cell_type)) {
    cell_type_de <-
        subset(combined_subset,
               subset = cell_type == {{ cell_type }}) %>%
        FindMarkers(ident.1 = "Control lung",
                    logfc.threshold = 0,
                    min.pct = 0) %>%
        as.data.frame() %>%
        rownames_to_column("gene") %>%
        mutate(cell_type = cell_type) %>%
        select(gene, avg_log2FC, p_val_adj, cell_type) %>%
        bind_rows(cell_type_de)
}

write_tsv(cell_type_de,
          file = "output/de/cell_type_de.txt")

for (factor_name in c("group_name", "sample_description")) {
    ############################################################################
    # Plot DE between lung and primary mets by cell type
    temp_data <-
        GetAssayData(combined_subset) %>%
        as.matrix() %>%
        t() %>%
        as.data.frame() %>%
        select(target_genes$mouse_gene[target_genes$mouse_gene %in%
                                    rownames(combined_subset)]) %>%
        rownames_to_column("cell") %>%
        mutate(cell_type = combined_subset$cell_type,
            group_name = combined_subset$group_name,
            sample_description = combined_subset$sample_description) %>%
        pivot_longer(cols = c(-group_name,
                             -cell_type,
                             -sample_description,
                             -cell),
                     names_to = "gene",
                     values_to = "expression")

    plot_name <-
        ggplot(temp_data,
               aes(x = gene,
                   y = expression,
                   color = get(factor_name))) +
        geom_point(size = 0.1,
                   alpha = 0.5,
                   position = position_jitterdodge()) +
        facet_wrap(~ cell_type, ncol = 1, scales = "free_y") +
        theme_roberts() +
        labs(x = "",
             y = "Expression",
             title = paste0("Gene Expression Split by ",
                            str_replace(factor_name, "_", " ") %>%
                                str_to_title(),
                            " and Cell Type")) +
        theme(legend.position = "right",
              legend.title = element_blank()) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(paste0("output/figures/jitter_cell_type_by_",
                  factor_name,
                  ".pdf"),
           plot = plot_name,
           width = 10,
           height = 8)

    plot_name <-
        temp_data %>%
        group_by(across(all_of(factor_name)),
                 cell_type,
                 gene) %>%
        summarize(prop_express = sum(expression > 0) / n(),
                  num_cells = n(),
                  .groups = "drop") %>%
        left_join(cell_type_de) %>%
        mutate(group = paste0(get(factor_name),
                            " (n=",
                            num_cells,
                            ")"),
            p_val_adj = replace_na(p_val_adj, 1),
            sig = ifelse(p_val_adj <= 0.05,
                            "Significant",
                            "Non-significant")) %>%
        ggplot(aes(x = group,
                   y = gene,
                   fill = prop_express)) +
        geom_tile(aes(color = sig),
                  size = 0.5,
                  width = 0.9,
                  height = 0.9) +
        geom_text(aes(label = sprintf("%.1f%%", prop_express * 100)),
                  size = 1,
                  color = "lightgray") +
        facet_wrap(~ cell_type, nrow = 1, scales = "free_x") +
        labs(x = "",
             y = "",
             title = paste0("Percent of Cells Expressing Select Gene")) +
        theme_roberts() +
        scale_fill_continuous(labels = scales::label_percent(accuracy = 2),
                              name = "Percent of\nCells Expressing Gene") +
        scale_color_manual(values = c(plot_cols[1], alpha("black", 0)),
                           name = NULL,
                           breaks = c("Significant", "Non-significant")) +
        theme(legend.position = "right",
              axis.text.x = element_text(angle = 90, hjust = 1)) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(file = paste0("output/figures/growth_factor_perc_expressing_",
                         factor_name,
                         ".pdf"),
        plot = plot_name,
        width = 10,
        height = 8)
}
```

Add primary: S0131


I'd like the following cell types to be in the main figure: dendritic cells, fibroblasts, macrophages, and monocytes And the following genes: AREG, EREG, FGF2, FGF10, FGF18, HBEGF, HGF, IFG1, NRG1, PDGFa, PDGFb, PDGFc, TGFa, TGFb1, and VEGFa
## Make main figures for growth factors
```{r main_figs, dependson='deGrowthFactorMets'}
combined_subset <- qs::qread("output/rdata/combined_subset.qs")

cell_type_de <- read_tsv("output/de/cell_type_de.txt",
                         show_col_types = FALSE)

main_figure_genes <-
    c("Areg", "Ereg", "Fgf2", "Fgf10", "Fgf18", "Hbegf", "Hgf", "Igf1", "Nrg1",
      "Pdgfa", "Pdgfb", "Pdgfc", "Tgfa", "Tgfb1", "Vegfa")

main_figure_cell_types <-
    c("Dendritic cells", "Fibroblasts", "Macrophages", "Monocytes")

main_jitter_cell_types <-
    c("Macrophages", "Monocytes")

combined_subset_main <-
    combined_subset %>%
    subset(cell_type %in% main_figure_cell_types)

combined_subset_main <-
    combined_subset_main[main_figure_genes, ]

for (factor_name in c("group_name", "sample_description")) {
    ############################################################################
    # Plot DE between lung and primary mets by cell type
    temp_data <-
        GetAssayData(combined_subset_main) %>%
        as.matrix() %>%
        t() %>%
        as.data.frame() %>%
        rownames_to_column("cell") %>%
        mutate(cell_type = combined_subset_main$cell_type,
               group_name = combined_subset_main$group_name,
               sample_description = combined_subset_main$sample_description) %>%
        pivot_longer(cols = c(-group_name,
                              -cell_type,
                              -sample_description,
                              -cell),
                     names_to = "gene",
                     values_to = "expression")

    plot_name <-
        temp_data %>%
        filter(cell_type %in% main_jitter_cell_types) %>%
        ggplot(aes(x = gene,
                   y = expression,
                   color = get(factor_name))) +
        geom_point(size = 0.1,
                   alpha = 0.5,
                   position = position_jitterdodge()) +
        facet_wrap(~ cell_type, ncol = 1, scales = "free_y") +
        theme_roberts() +
        labs(x = "",
             y = "Expression",
             title = paste0("Gene Expression Split by ",
                            str_replace(factor_name, "_", " ") %>%
                                str_to_title(),
                            " and Cell Type")) +
        theme(legend.position = "right",
              legend.title = element_blank()) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(paste0("output/figures/jitter_cell_type_by_",
                  factor_name,
                  "_main.pdf"),
           plot = plot_name,
           width = 8,
           height = 4)

    plot_name <-
        temp_data %>%
        group_by(across(all_of(factor_name)),
                 cell_type,
                 gene) %>%
        summarize(prop_express = sum(expression > 0) / n(),
                  num_cells = n(),
                  .groups = "drop") %>%
        left_join(cell_type_de) %>%
        mutate(group = paste0(get(factor_name),
                            " (n=",
                            num_cells,
                            ")"),
            p_val_adj = replace_na(p_val_adj, 1),
            sig = ifelse(p_val_adj <= 0.05,
                            "Significant",
                            "Non-significant")) %>%
        ggplot(aes(x = group,
                   y = gene,
                   fill = prop_express)) +
        geom_tile(aes(color = sig),
                  size = 1,
                  width = 0.9,
                  height = 0.9) +
        geom_text(aes(label = sprintf("%.1f%%", prop_express * 100)),
                  size = 2,
                  color = "lightgray") +
        facet_wrap(~ cell_type, nrow = 1, scales = "free_x") +
        labs(x = "",
             y = "",
             title = paste0("Percent of Cells Expressing Select Gene")) +
        theme_roberts() +
        scale_fill_continuous(labels = scales::label_percent(accuracy = 2),
                              name = "Percent of\nCells Expressing Gene") +
        scale_color_manual(values = c(plot_cols[1], alpha("black", 0)),
                           name = NULL,
                           breaks = c("Significant", "Non-significant")) +
        theme(legend.position = "right",
              axis.text.x = element_text(angle = 90, hjust = 1)) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(file = paste0("output/figures/growth_factor_perc_expressing_",
                         factor_name,
                         "_main.pdf"),
        plot = plot_name,
        width = 8,
        height = 6)
}
```


```{r, cache=FALSE}
sessionInfo()
```


## Human Down Sampling for human
```{r}
#load the sobj
load("output/rdata/sobj_list.rds")
#subset human only
human_objects_list <- sobj_list[c("tibia_tumor_h",
                                "d14_pos_h",
                                "d28_pos_h",
                                "d49_pos_h" )]

human_objects_list

#Downsample
for (i in seq_len(length(human_objects_list))) {
    temp_obj <- human_objects_list[[i]]
    current_n_cells <- length(Cells(temp_obj))
    sub_n_cells <- min(1500, current_n_cells)
    temp_obj <- temp_obj[, sample(1:current_n_cells, sub_n_cells)]
    human_objects_list[[i]] <- temp_obj
}

human_objects_list

#Dimplot with down sample
downsample_umap_human_only <-
    merge(human_objects_list[["tibia_tumor_h"]],
            y = human_objects_list[c("d14_pos_h",
                            "d28_pos_h",
                            "d49_pos_h")],
            add.cell.ids = c("Primary_tumor",
                            "Early_metastatic_tumor",
                            "Midpoint_metastatic_tumor",
                            "Late_metastatic_tumor"),
            project = "downsample_umap_human_only") %>%
            process_seurat()

#Dimplot
downsample_human_combined <-
    DimPlot(downsample_umap_human_only) +
    ggtitle(label="Primary and metastatic tumor") +
    theme_roberts() +
    theme(plot.title = element_text(hjust = 0.5)) +
    coord_fixed()
downsample_human_combined

#split by
downsample_human_split <-
    DimPlot(downsample_umap_human_only, split.by = "group") +
            theme_roberts() +
            theme(plot.title = element_text(hjust = 0.5)) +
            NoLegend() +
            coord_fixed()
downsample_human_split

#rearrange the panels for umap space
downsample_human_split$data$group <-
    factor(x = downsample_human_split$data$group,
           levels = c("Primary_tumor",
                      "Early_metastatic_tumor",
                      "Midpoint_metastatic_tumor",
                      "Late_metastatic_tumor"))

downsample_human_split

#save the plot
ggsave("output/figures/downsample_human_combined.pdf",
       width = 6,
       height = 6,
       plot = downsample_human_combined)

ggsave("output/figures/downsample_human_split.pdf",
       width = 6,
       height = 6,
       plot = downsample_human_split)

```


## Mouse Down Sampling and Dimplot
```{r}
#subset human only
mouse_objects_list <- sobj_list[c("S0006_mouse",
                                "d14_pos_m",
                                "d28_pos_m",
                                "d49_pos_m")]

mouse_objects_list

#Downsample
for (i in seq_len(length(mouse_objects_list))) {
    temp_obj <- mouse_objects_list[[i]]
    current_n_cells <- length(Cells(temp_obj))
    sub_n_cells <- min(4500, current_n_cells)
    temp_obj <- temp_obj[, sample(1:current_n_cells, sub_n_cells)]
    mouse_objects_list[[i]] <- temp_obj
}
mouse_objects_list

#downsampled objects are merged and seurat process for dimplot
downsample_umap_mouse_only <-
    merge(mouse_objects_list[["S0006_mouse"]],
            y = mouse_objects_list[c("d14_pos_m",
                            "d28_pos_m",
                            "d49_pos_m")],
            add.cell.ids = c("Mouse_control_lung",
                            "Early_mouse_stroma",
                            "Midpoint_mouse_stroma",
                            "Late_mouse_stroma"),
            project = "downsample_umap_mouse_only") %>%
            process_seurat()

#combined
downsample_mouse_combined <- DimPlot(downsample_umap_mouse_only) +
                        theme_roberts() +
                        labs(title = "Mouse control and stroma") +
                        theme(plot.title = element_text(hjust = 0.5)) + 
                        coord_fixed()
downsample_mouse_combined

#split by
downsample_mouse_split <- 
            DimPlot(downsample_umap_mouse_only, split.by = "group") +
            theme_roberts() +
            theme(plot.title = element_text(hjust = 0.5)) +
            NoLegend() + coord_fixed()
downsample_mouse_split


#rearrange the panels for umap space
downsample_mouse_split$data$group <- 
    factor(x = downsample_mouse_split$data$group,
           levels = c("Mouse_control_lung",
                      "Early_mouse_stroma",
                       "Midpoint_mouse_stroma",
                       "Late_mouse_stroma"))

downsample_mouse_split

#save the plot
ggsave("output/figures/downsample_mouse_combined.pdf",
       width = 6,
       height = 6,
       plot = downsample_mouse_combined)

ggsave("output/figures/downsample_mouse_split.pdf",
       width = 6,
       height = 6,
       plot = downsample_mouse_split)


#all combined

plots_all <-
    downsample_human_combined + downsample_human_split +
    downsample_mouse_combined + downsample_mouse_split +
    plot_layout(widths = c(1, 4))
plots_all


#save the plot
ggsave("output/figures/plots_all.pdf",
       width = 10,
       height = 4,
       plot = plots_all)

```



## Non-downsample UMAP plots
```{r}
#load the sobj
load("output/rdata/sobj_list.rds")

#subset the sobj of interest for human and mouse and merge
#for human
umap_human_only <-
    merge(sobj_list[["tibia_tumor_h"]],
          y = sobj_list[c("d14_pos_h",
                            "d28_pos_h",
                            "d49_pos_h")],
          add.cell.ids = c("Primary_tumor",
                            "Early_metastatic_tumor",
                            "Midpoint_metastatic_tumor",
                            "Late_metastatic_tumor"),
           project = "umap_human_only") %>%
    process_seurat()

#Human only combined umap: this puts them in the same space/comparable
combined_human_only <- 
    DimPlot(umap_human_only) +
    ggtitle(label = "Primary and metastatic tumor") +
    theme_roberts() +
    theme(plot.title = element_text(hjust = 0.5)) +
    coord_fixed()

combined_human_only

#Dimplot
human_umaps <- DimPlot(umap_human_only, split.by = "group") +
            theme_roberts() +
            theme(plot.title = element_text(hjust = 0.5)) +
            NoLegend() + coord_fixed()

human_umaps

#rearrange the panels for umap space
human_umaps$data$group <- factor(x = human_umaps$data$group,
                                levels = c("Primary_tumor",
                                        "Early_metastatic_tumor",
                                        "Midpoint_metastatic_tumor",
                                        "Late_metastatic_tumor"))
human_umaps

#for mouse
umap_mouse_only <-
    merge(sobj_list[["S0006_mouse"]],
            y = sobj_list[c("d14_pos_m",
                            "d28_pos_m",
                            "d49_pos_m")],
            add.cell.ids = c("Mouse_control_lung",
                            "Early_mouse_stroma",
                            "Midpoint_mouse_stroma",
                            "Late_mouse_stroma"),
            project = "umap_mouse_only") %>%
            process_seurat()


#Dimplot
mouse_umaps_combined <- DimPlot(umap_mouse_only) +
                        theme_roberts() +
                        theme(plot.title = element_text(hjust = 0.5)) +
                        NoLegend() + 
                        coord_fixed()

mouse_umaps <- DimPlot(umap_mouse_only, split.by = "group") +
               theme_roberts() +
               theme(plot.title = element_text(hjust = 0.5)) +
               NoLegend() + 
               coord_fixed()

mouse_umaps

#rearrange the panels for umap space
mouse_umaps$data$group <- factor(x = mouse_umaps$data$group,
                                levels = c("Mouse_control_lung",
                                        "Early_mouse_stroma",
                                        "Midpoint_mouse_stroma",
                                        "Late_mouse_stroma"))

mouse_umaps

combined_human_only

combined_plots <- (human_umaps / mouse_umaps)
combined_plots

combined_plots_all <- (combined_human_only + (human_umaps / mouse_umaps)) 
                         + plot_layout(widths = c(1, 4))
combined_plots_all


#save the plot
ggsave("output/figures/combined_human_only.pdf",
       width = 6,
       height = 6,
       plot = combined_plots)

ggsave("output/figures/combined_plots.pdf",
       width = 8,
       height = 6,
       plot = combined_plots)

ggsave("output/figures/combined_plots_all.pdf",
       width = 12,
       height = 6,
       plot = combined_plots)


```


## GSEA_KEGG
GSEA-KEGG figure showing upregulated genes/pathways as lesions progress from: 
                  S0037 (primary) redo human cells to S0082(early) human cells 
                  S0082 (early) human cell to S0084(midpoint) human cells 
                  S0084(midpoint) human cells to S0093 (late) human cells 
                  S0082(early) human cells to S0093(late) human cells 
```{r}
load("output/rdata/sobj_list.rds")
#we want all of the item in the same space, do combine and run process seurat
human_only <-
    merge(sobj_list[["tibia_tumor_h"]],
          y = sobj_list[c("d14_pos_h",
                          "d28_pos_h",
                          "d49_pos_h")],
          add.cell.ids = c("Primary_tumor",
                           "Early_metastatic_tumor",
                           "Midpoint_metastatic_tumor",
                           "Late_metastatic_tumor"),
          project      =   "human_only") %>%
          process_seurat()

#set the idents
Idents(human_only) <- "group"

#make a table to do the find markers in loop
marker_table <- tribble(~item1, ~item2,
                        "Primary_tumor", "Early_metastatic_tumor",
                        "Early_metastatic_tumor", "Midpoint_metastatic_tumor",
                        "Midpoint_metastatic_tumor", "Late_metastatic_tumor",
                        "Late_metastatic_tumor", "Primary_tumor")

#make a list that you can append to
markers_list <- list()
gsea_gene_lists <- list()
#FindMarkers since we comaring across only two groups
for (i in 1:nrow(marker_table)) {
        df <- FindMarkers(human_only,
                         ident.1 = marker_table$item1[i],
                         ident.2 = marker_table$item2[i],
                         min.pct = 0.1,
                         logfc.threshold = 0)
        df$gene <- rownames(df) 
        gsea_list <- df %>% 
              arrange(desc(avg_log2FC)) %>%
              dplyr::select(gene, avg_log2FC)
        gsea_list <- deframe(gsea_list)
        filename1 <- paste0(marker_table$item1[i],
                            "_",
                            marker_table$item2[i],
                            ".tsv")
        filename2 <- paste0(marker_table$item1[i],
                            "_vs_",
                            marker_table$item2[i],
                            ".tsv")
        write_tsv(as.data.frame(df), file = filename1)
        write_tsv(as.data.frame(gsea_list), file = filename2)
        markers_list[[paste0(marker_table$item1[i],
                            "_",
                            marker_table$item2[i])]] <- df
        gsea_gene_lists[[paste0(marker_table$item1[i],
                            "_vs_",
                            marker_table$item2[i])]] <- gsea_list
}



# Running the GSEA_KEGG
# select dataset and then aggregate the genes into a list
kegg_ref <- msigdbr::msigdbr(species = "Homo sapiens",
                category = "C2",
                subcategory = "CP:KEGG") %>%
                split(x = .$gene_symbol, f = .$gs_name)


#Run the GSEA in loop for all the things
kegg_results_list <- list()
top5up_down <- list()
for (item in names(gsea_gene_lists)) {
    print(item)
    result <- fgsea::fgseaMultilevel(kegg_ref,
                            gsea_gene_lists[[item]],
                            minSize = 15,      # min size of gene set to test
                            maxSize = 500,     # max size of a gene set to test
                            nPerm = 1000) %>%  # number of permutations to test
                            arrange(padj)
    result <- result %>% arrange(desc(NES))
    top5up <- result %>% slice_head(n = 5)
    top5down <- result %>% slice_tail(n = 5)
    temp_top5 <- rbind(top5up, top5down)
    top5up_down[[item]] <- temp_top5 %>% 
                             dplyr::select(pathway, NES)
    filename <- paste0(markers_list[item],".tsv")
    write_tsv(as.data.frame(result), file = filename)
    kegg_results_list[[item]] <- result
}


#Visualize the top 5 up and downregulated pathways
primary_vs_early <-
    top5up_down$Primary_tumor_vs_Early_metastatic_tumor
early_vs_midpoint <-
    top5up_down$Early_metastatic_tumor_vs_Midpoint_metastatic_tumor



#Use loop to do what
#for (item in top5up_down) {
    #print(item)
    #for (stuff in item$pathway) {
        #print(stuff)
        #for (i in stuff) {
            #stuff <- stringr::str_replace_all(stuff, "_", " ")
        #}
    #}
#}


#Replace the "_" with space
primary_vs_early$pathway <-
    stringr::str_replace_all(primary_vs_early$pathway, "_", " ")

#Remove KEGG word
primary_vs_early$pathway <-
    stringr::str_replace_all(primary_vs_early$pathway, "KEGG", "")

#Pass the Pathways name as Sentence
primary_vs_early$pathway <-
    stringr::str_to_sentence(primary_vs_early$pathway)

#prepare the dataframe for the plot 
df <- data.frame(
    label_y = c(1, 1, 1, 1, 1, -1, -1, -1, -1, -1),
    order = c(1, 2, 3, 4, 5, 5, 4, 3, 2, 1))

#Practice in one of them
primary_vs_early <- cbind(primary_vs_early, df)

df$tissue = "Primary vs Early Geneset"


#prepare the label for the plot
lab4plot <- tibble(y = c(-4, 4),
                   x = c(0.5, 0.5),
                   label = as.factor(c("Upregulated in Primary Tumor",
                                      "Upregulated in Early Metastatic Tumor")),
                   levels = c("Primary vs Early Geneset"))

## Ran IPA on the output of FindMarkers above
Used a logFC cutoff of -1/1 and p_val_adj cutoff of 0.05
```{r plotIPA}
path_top_5 <-
    read_tsv(list.files("output/pathway_analysis",
                        pattern = "_paths.txt",
                        full.names = TRUE),
         id = "sample",
         col_names = c("pathway",
                       "neg_log_pval",
                       "ratio",
                       "z_score",
                       "genes"),
         show_col_types = FALSE,
         skip = 1) %>%
    filter(neg_log_pval > log10(0.05) &
           !is.nan(z_score) &
           z_score != 0) %>%
    mutate(sample = str_remove(sample, ".+/") %>%
               str_remove("_paths.txt")) %>%
    group_by(sample, z_score > 0) %>%
    arrange(desc(abs(z_score)), .by_group = TRUE) %>%
    slice_head(n = 5) %>%
    select(sample, pathway, z_score) %>%
    mutate(order = seq_len(n()),
           label_y = z_score / abs(z_score),
           sample = str_remove(sample, "_tumor$") %>%
                    str_replace_all("metastatic", "metastasis") %>%
                    str_replace("tumor", "vs") %>%
                    str_replace_all("_",
                                    " ") %>%
                    str_to_sentence() %>%
                    factor(., levels = c("Primary vs early metastasis",
                                         "Early metastasis vs midpoint metastasis",
                                         "Midpoint metastasis vs late metastasis",
                                         "Late metastasis vs primary"))) %>%
    ungroup()

two_way_plot <- function(data) {
    lab4plot <-
        tibble(y = c(-3, 3),
               x = c(0.2, 0.2),
               label = as.factor(c(paste("Upregulated in",
                                         str_remove(data$sample[1],
                                                    ".+vs") %>%
                                            str_to_lower()),
                                   paste("Upregulated in",
                                         str_remove(data$sample[1],
                                                    "vs.+") %>%
                                            str_to_lower()))))

    plot_name <-
        ggplot() +
        geom_bar(data = data,
                 aes(x = -1 * order,
                     y = z_score,
                     fill = z_score > 0),
                 stat = "identity",
                 alpha = 0.8) +
            coord_flip() +
            geom_hline(yintercept = 0,
                       color = "black",
                       linewidth = 0.5) +
            geom_text(data = data,
                      aes(x = -1 * order,
                          y = label_y * 3,
                          label = pathway),
                      size = 1.5) +
            geom_text(data = lab4plot,
                      aes(x = x,
                          y = y,
                          label = label),
                      fontface = "bold",
                      size = 1.5) +
            scale_fill_manual(values = plot_cols) +
            # theme_roberts(title_font_size = 5,
            #               axis_font_size = 4) +
            theme(strip.background = element_rect(color = "white",
                                                  fill = "white"),
                  legend.position = "none",
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank(),
                  panel.border = element_blank(),
                  axis.line.x = element_line(color = "black"),
                  axis.line.y = element_blank(),
                  axis.text.y = element_blank(),
                  axis.ticks.y = element_blank(),
                  plot.title = element_text(size = 5, face = "bold"),
                  axis.text.x = element_text(size = 3)) +
            labs(title = data$sample[1],
                 y = "",
                 x = "") +
            ylim(-5, 5)

    return(plot_name)
}

plot_list <- lapply(levels(path_top_5$sample),
                    function(x) {
    two_way_plot(filter(path_top_5, sample == x))
})

pdf("output/figures/Top_5_IPA_pathways.pdf",
    width = 4,
    height = 4)
wrap_plots(plot_list, ncol = 1, heights = c(5, 5, 1.8, 3))
dev.off()


# Save as PDF
pdf("/home/gdrobertslab/lab/Analysis/Yogesh/MCL1_early_met_survival/Figures/figure_2c.pdf",
    width = 6,
    height = 6)

figure_2c
dev.off()

save(figure_2c, file = "Figures/figure_2c.pdf.RData")


# make bubble plot
bubble <- ggplot(top5up_down$Primary_tumor_vs_Early_metastatic_tumor,
                aes(x = NES,
                    y = pathway,
                    size = size,
                    color = padj)) +
                geom_point() +
                scale_color_gradient2(high = "#ff0000",
                    mid = "#ffffff",
                    low = "#3942c0",
                    midpoint = 0.05)
bubble


```



# Nichenetr
sender : all stroma/mouse cells from the early+mid
receiver : Human cells/tumor from the early+mid 
Genes/markers tested : differential expression from ident1=early and ident2=mid
```{r}  
load("output/rdata/sobj_list.rds")

#seurat objects sender of signal
sender_sobj <- 
    merge(sobj_list[["d14_pos_m"]],
          y = sobj_list[c("d28_pos_m")],
          add.cell.ids = c("Early_mouse_stroma",
                            "Midpoint_mouse_stroma"),
            project = "sender_sobj") %>%
            process_seurat()

#seurat objects receiver of signal
receiver_sobj <- 
    merge(sobj_list[["d14_pos_h"]],
          y = sobj_list[c("d28_pos_h")],
          add.cell.ids = c("Early_metastatic_tumor",
                           "Midpoint_metastatic_tumor"),
          project      =   "receiver_sobj") %>%
          process_seurat()


```

## Get the pre-defined model data
```{r}

ligand_target_matrix <-
    readRDS( "nichenetrfiles/ligand_target_matrix.rds")

lr_network <-
    readRDS("nichenetrfiles/lr_network.rds")

weighted_networks <-
    readRDS("nichenetrfiles/weighted_networks.rds")

weighted_networks_lr <-
    weighted_networks$lr_sig %>%
    inner_join(lr_network %>% distinct(from, to),
               by = c("from", "to"))


```


## Convert the gene names in model to mouse
```{r}
#makes capital lettered human genes (CXCL) to mouse format (Cxcl)
lr_network <-
    lr_network %>%
    mutate(from = convert_human_to_mouse_symbols(from),
           to = convert_human_to_mouse_symbols(to)) %>%
    drop_na()

#converts the colnames human genes name format to mouse format
colnames(ligand_target_matrix) <-
    ligand_target_matrix %>%
    colnames() %>%
    convert_human_to_mouse_symbols()

#converts the rownames human genes name format to mouse format
rownames(ligand_target_matrix) <-
    ligand_target_matrix %>%
    rownames() %>%
    convert_human_to_mouse_symbols()

#removes the NAs or the genes that were umable to convert
ligand_target_matrix <-
    ligand_target_matrix %>%
    .[!is.na(rownames(ligand_target_matrix)),
        !is.na(colnames(ligand_target_matrix))]

#Convert the human format genes to mouse format in the file and drops NAs
weighted_networks_lr <-
    weighted_networks_lr %>%
    mutate(from = convert_human_to_mouse_symbols(from),
           to = convert_human_to_mouse_symbols(to)) %>%
    drop_na()

```





