---
title: "Data Analysis for Early and Late Lung-colonizing Tumors"
author: "Emily Franz and Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code_folding: hide
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    echo = TRUE,
    cache = TRUE,
    collapse = TRUE,
    tidy.opts = list(width.cutoff = 95),
    message = FALSE,
    warning = FALSE,
    cache.lazy = FALSE
)
```

```{r lib}
# Load packages
library(Seurat)
library(tidyverse)
library(rrrSingleCellUtils)
theme_set(theme_bw())
theme_update(plot.title = element_text(hjust = 0.5))
library(ggplot2)
library(patchwork)

# Set random generator seed to facilitate reproducibility
set.seed(888)
```

## Make up directory structure
```{bash mkdirs, eval=TRUE}
for directoryName in \
  output \
  output/figures \
  output/rdata \
  output/de
do
  if [ ! -d ${directoryName} ]
  then
    mkdir ${directoryName}
  fi
done
```

# Load and Process Datasets
Datasets are from tumors collected from mouse lungs at early (day 14 post injection of tumor cells) and late (day 35 post injection of tumor cells) timepoints. mCherry was used to label the tumor cells as well as the surrounding niche cells.
```{r load}
sample_list <- read_tsv("misc/sample_cutoffs.txt", show_col_types = FALSE)

sobj_list <- list()

for (i in seq_len(nrow(sample_list))) {
    species_pattern <- sample_list$species_pattern[i]
    obj_name <- sample_list$obj_name[i]

    sobj_list[[obj_name]] <-
        tenx_load_qc(paste0("/home/gdrobertslab/lab/Counts/",
                            sample_list$sample_name[i],
                            "/filtered_feature_bc_matrix"),
                     violin_plot = FALSE,
                     species_pattern = species_pattern)

    # Add metadata to dataset
    for (colname in colnames(sample_list)) {
        sobj_list[[obj_name]][[colname]] <- sample_list[[colname]][i]
    }

    sobj_list[[obj_name]] <-
        sobj_list[[obj_name]] %>%
        subset(nCount_RNA >= sample_list$ncount_rna_min[i] &
               nCount_RNA <= sample_list$ncount_rna_max[i] &
               percent.mt <= sample_list$mt_max[i]) %>%
        process_seurat()

    plotted <-
        feature_hist(sobj_list[[obj_name]],
                     features = c("nCount_RNA", "percent.mt"))

    ggsave(paste0("output/figures/feature_hist_",
                  obj_name,
                  ".png"),
           width = 10,
           height = 10,
           plot = plotted)
}
save(sobj_list,
     file = "output/rdata/sobj_list.rds")
```

## UMAP with all samples
```{r umap_alldependson='load', cache.vars=''}
load("output/rdata/sobj_list.rds")
all_samples <-
    merge(sobj_list[[1]],
          y = sobj_list[2:length(sobj_list)],
          add.cell.ids = names(sobj_list)) %>%
    process_seurat(resolution = 0.2)

plot_name <-
    DimPlot(all_samples,
            group.by = "sample_description",
            split.by = "sample_description",
            shuffle = TRUE,
            ncol = 2,
            cols = plot_cols[seq_along(sobj_list)],
            pt.size = 0.01,
            order = FALSE) +
    theme_roberts() +
    labs(title = NULL)

ggsave("output/figures/umap_all_samples.pdf",
       width = 4,
       height = 6,
       plot = plot_name)
```

## Human Cell Only Processing
```{r human-processing, dependson="load"}
# Human
# Normalize the number of cells per sample
n_cells <- c()

for (i in seq_along(sobj_list)) {
    n_cells[i] <- length(colnames(sobj_list[[i]]))
}

names(n_cells) <- names(sobj_list)

# Merge day 14 and day 28 Seurat objects together
tumor <-
    merge(sobj_list[["d14_pos_h"]],
          y = sobj_list[c("d28_pos_h", "tibia_tumor_h")],
          add.cell.ids = c("d14", "d28", "tibia"),
          project = "Early vs Late OS Mets") %>%
    NormalizeData() %>%
    ScaleData() %>%
    FindVariableFeatures() %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    FindClusters(resolution = 0.15, k.param = 30) %>%
    RunUMAP(dims = 1:30, n.neighbors = 30L, metric = "euclidean")

# Save human mCherry+ merged object
save(tumor, file = "output/rdata/d14_d28_tumor.RData")

# Use timepoint naming (Day 14/Day 28) for UMAP plotting
umap_timepoint <-
    r_dim_plot(tumor,
        title = "Tumor at metastatic timepoints",
        group.by = "timepoint",
        label = TRUE)
umap_timepoint

pdf("output/figures/d14_d28_umap.pdf",
    width = 5,
    height = 5)
umap_timepoint
dev.off()

# Use basic naming (early/late) for UMAP plotting
umap_group <-
    r_dim_plot(tumor,
        title = "Tumor at metastatic timepoints",
        group.by = "group",
        label = TRUE,
        repel = TRUE)
umap_group

pdf("output/figures/d14_d28_umap.pdf",
    width = 5,
    height = 5)
umap_group
dev.off()

# Reorder the group metadata variable
tumor$group <- factor(tumor$group, levels = c("Primary", "Early", "Late"))

umap_group_split <-
    r_dim_plot(tumor,
        title = "Tumor at metastatic timepoints",
        split.by = "group",
        label = TRUE,
        repel = TRUE)
umap_group_split

```

## Mouse Cell Only Processing

```{r mouse-processing, dependson="load"}
# Mouse
# Merge day 14 and day 35 Seurat objects together
microenv <-
    merge(sobj_list[["d14_pos_m"]],
          y = sobj_list[["d35_pos_m"]],
          add.cell.ids = c("d14", "d35"),
          project = "Early vs Late OS Mets")

# Process merged object
microenv <-
    NormalizeData(microenv) %>%
    ScaleData() %>%
    FindVariableFeatures() %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    FindClusters(resolution = 0.15, k.param = 30) %>%
    RunUMAP(dims = 1:30, n.neighbors = 30L, metric = "euclidean")

# Plot UMAP
env_umap <- r_dim_plot(microenv,
        title = "Stroma at metastatic timepoints",
        label = TRUE)
env_umap

env_umap_split <- r_dim_plot(microenv,
        title = "Stroma at metastatic timepoints",
        split.by = "group")
env_umap_split

timepoints_panel <- (umap_group | umap_group_split) / (env_umap | env_umap_split)
timepoints_panel

# Save mouse mCherry+ merged Seurat object
save(microenv, file = "output/rdata/early_late_microenv.RData")
```

# Plotting Expression: Genes of Interest

## Human Cells Only: Pre-cell cycle regression
```{r human_geneplots, dependson=c("load","human-processing")}
# Dot plot of expression of genes of interest by group (early/late)
gene_list <-
    c("BCL2",
      "BCL2L1",
      "BCL2L2",
      "MCL1")

dot_plot <-
    DotPlot(tumor,
            features = gene_list,
            group.by = "group",
            scale = FALSE,
            cols = c("blue", "red")) +
  RotatedAxis() +
  coord_fixed()

# Save as pdf
pdf("output/figures/early_late_dot_plot.pdf",
    width = 7,
    height = 4)
dot_plot
dev.off()

# Violin plot of expression of genes of interest by group (early/late)
vln_plot <-
    VlnPlot(tumor,
            features = gene_list,
            group.by = "group") +
    RotatedAxis()

# Save as pdf
pdf("output/figures/early_late_vln_plot.pdf",
    width = 6,
    height = 5)
vln_plot
dev.off()

# Feature plot (UMAP) of expression of genes of interest by group (early/late)
feature_plot <-
    FeaturePlot(tumor,
                features = gene_list,
                split.by = "group",
                order = T)

# Save as pdf
pdf("output/figures/early_late_feature_plot.pdf",
    width = 6,
    height = 12)
feature_plot
dev.off()
```

## Human Cells Only: Post-cell cycle regression

Regress out cell cycle genes so that the dataset does not cluster on the UMAP 
by cell cycle gene expression.

```{r cc, dependson=c("load","human-processing")}
# Calculate cell cycle phase and view its effects on clustering
# Note: default for kill_cc() is not performing cell cycle regression

# Save as pdf
pdf("output/figures/early_late_tumor_pre-kill_cc.pdf",
    width = 5,
    height = 5)
kill_cc(tumor)
dev.off()

# Calculate cell cycle phase and remove its effects on clustering
tumor_cc <-
    kill_cc(tumor,
            cc_regress = "Y") %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.15) %>%
    RunUMAP(dims = 1:20)

# Plot tumor UMAP after cell cycle genes are regressed out
DimPlot(tumor_cc,
        group.by = "Phase",
        label = TRUE,
        pt.size = 1) +
    theme_roberts()

# Save as pdf
pdf("output/figures/early_late_post-kill_cc.pdf",
    width = 5,
    height = 5)
DimPlot(tumor_cc,
        group.by = "Phase",
        label = TRUE,
        pt.size = 1) +
    theme_roberts()
dev.off()

# Plot tumor UMAP by timepoint post cell cycle regression
umap_cc <-
    DimPlot(tumor_cc,
            pt.size = 1,
            label = TRUE,
            group.by = "timepoint") +
    theme_roberts() +
    theme(aspect.ratio = 1)
umap_cc

# Save as pdf
pdf("output/figures/early_late_umap_cc_bytrtment.pdf",
    width = 5,
    height = 5)
umap_cc
dev.off()

# Save the cell cycle regressed tumor object
save(tumor_cc, file = "output/rdata/early_late_tumor_cc.RData")
```

Repeat plots of gene expression post cell cycle regression.
Add S0131 to dotplot
```{r human_geneplots_postcc, dependson=c("load","human-processing","cc")}
# Dot plot of expression of genes of interest by group (early/late)
dot_plot <-
    DotPlot(tumor_cc,
            features = gene_list,
            group.by = "group",
            scale = F,
            cols = c("blue", "red")) +
  RotatedAxis() +
  coord_fixed()

# Save as pdf
pdf("output/figures/early_late_dot_plot_postccr.pdf",
    width = 7,
    height = 4)
dot_plot
dev.off()

# Violin plot of expression of genes of interest by group (early/late)
vln_plot <-
    VlnPlot(tumor_cc,
            features = gene_list,
            group.by = "group") +
  RotatedAxis()

# Save as pdf
pdf("output/figures/early_late_vln_plot_postccr.pdf",
    width = 6,
    height = 5)
vln_plot
dev.off()

# Feature plot (UMAP) of expression of genes of interest by group (early/late)
feature_plot <-
    FeaturePlot(tumor_cc,
                features = gene_list,
                split.by = "group",
                order = T) +
  theme(aspect.ratio = 1)

# Save as pdf
pdf("output/figures/early_late_feature_plot_postccr.pdf",
    width = 6,
    height = 12)
feature_plot
dev.off()
```

## Growth factor expression in normal vs lungs with mets
Also plot each time point separately as a time-course
```{r deGrowthFactorMets, dependson='load'}
cm_target_genes <-
    read_tsv("misc/growth_factor_genes.txt",
             col_names = "human_gene",
             show_col_types = FALSE) %>%
    left_join(read_tsv("misc/HMD_HumanPhenotype.rpt",
                       col_names = c("human_gene",
                                     "h_entrez",
                                     "mouse_gene",
                                     "mgi",
                                     "pheno",
                                     "junk"),
                       show_col_types = FALSE) %>%
        select(human_gene, mouse_gene),
        by = "human_gene") %>%
    select(mouse_gene)

load("output/rdata/sobj_list.rds")

samples <- c("S0006_mouse", "d14_pos_m", "d28_pos_m", "d49_pos_m")

sobj_list$S0006_mouse$group_name <- "Control lung"

for (sample_name in c("d14_pos_m", "d28_pos_m", "d49_pos_m")) {
    sobj_list[[sample_name]]$group_name <- "Met. lung"
}

combined_data <-
    merge(x = sobj_list$S0006_mouse,
          y = list(sobj_list[["d14_pos_m"]],
                   sobj_list[["d28_pos_m"]],
                   sobj_list[["d49_pos_m"]])) %>%
    process_seurat()

combined_data$sample_description <-
    factor(combined_data$sample_description,
           levels = c("SCID control lung",
                      "OS17 mets - 2wk",
                      "OS17 mets - 4wk",
                      "OS17 mets - 7wk"))

combined_data$sample_name <-
    combined_data$sample_name %>%
    str_remove("/outs") %>%
    factor(., levels = c("S0006-WTLung",
                         "S0082",
                         "S0084",
                         "S0093"))
```

### Run DE on the tumor samples
```{r deGrowthFactorMetsDE, dependson='deGrowthFactorMets'}
Idents(combined_data) <- combined_data$group_name

overall_de <-
    FindMarkers(combined_data,
                ident.1 = "Control lung",
                logfc.threshold = 0)
```

### Assign cell types
```{r deGrowthFactorMetsCellType, dependson='deGrowthFactorMetsDE'}
imm_cells <- celldex::ImmGenData()
mouse_rna <- celldex::MouseRNAseqData()

cell_assign <-
    SingleR::SingleR(as.SingleCellExperiment(combined_data),
                     ref = list(imm_cells,
                                mouse_rna),
                     labels = list(imm_cells$label.main,
                                   mouse_rna$label.main))

combined_data$cell_type <-
    cell_assign$labels %>%
    str_replace("DC", "Dendritic cells")
combined_data$cell_score <-
    cell_assign$scores %>%
    apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

cell_type_list <-
    table(combined_data$cell_type, combined_data$group) %>%
    as.data.frame() %>%
    group_by(Var1) %>%
    summarize(sum = sum(Freq), .groups = "drop") %>%
    filter(sum > 50)
```

### Make supplemental figures with all cell types
```{r deGrowthFactorMetsLoop, dependson='deGrowthFactorMetsCellType'}
target_genes <- cm_target_genes

target_genes <-
    left_join(target_genes,
              overall_de %>%
              rownames_to_column("mouse_gene"))

################################################################################
# DE between lung and primary mets by cell type
combined_subset <-
    combined_data %>%
    subset(cell_type %in% cell_type_list$Var1)

combined_subset <-
    combined_subset[rownames(combined_subset) %in% target_genes$mouse_gene, ]

Idents(combined_subset) <- combined_subset$group_name

qs::qsave(combined_subset,
          file = "output/rdata/combined_subset.qs")

cell_type_de <- tibble(gene = character())

for (cell_type in unique(combined_subset$cell_type)) {
    cell_type_de <-
        subset(combined_subset,
               subset = cell_type == {{ cell_type }}) %>%
        FindMarkers(ident.1 = "Control lung",
                    logfc.threshold = 0,
                    min.pct = 0) %>%
        as.data.frame() %>%
        rownames_to_column("gene") %>%
        mutate(cell_type = cell_type) %>%
        select(gene, avg_log2FC, p_val_adj, cell_type) %>%
        bind_rows(cell_type_de)
}

write_tsv(cell_type_de,
          file = "output/de/cell_type_de.txt")

for (factor_name in c("group_name", "sample_description")) {
    ############################################################################
    # Plot DE between lung and primary mets by cell type
    temp_data <-
        GetAssayData(combined_subset) %>%
        as.matrix() %>%
        t() %>%
        as.data.frame() %>%
        select(target_genes$mouse_gene[target_genes$mouse_gene %in%
                                    rownames(combined_subset)]) %>%
        rownames_to_column("cell") %>%
        mutate(cell_type = combined_subset$cell_type,
            group_name = combined_subset$group_name,
            sample_description = combined_subset$sample_description) %>%
        pivot_longer(cols = c(-group_name,
                             -cell_type,
                             -sample_description,
                             -cell),
                     names_to = "gene",
                     values_to = "expression")

    plot_name <-
        ggplot(temp_data,
               aes(x = gene,
                   y = expression,
                   color = get(factor_name))) +
        geom_point(size = 0.1,
                   alpha = 0.5,
                   position = position_jitterdodge()) +
        facet_wrap(~ cell_type, ncol = 1, scales = "free_y") +
        theme_roberts() +
        labs(x = "",
             y = "Expression",
             title = paste0("Gene Expression Split by ",
                            str_replace(factor_name, "_", " ") %>%
                                str_to_title(),
                            " and Cell Type")) +
        theme(legend.position = "right",
              legend.title = element_blank()) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(paste0("output/figures/jitter_cell_type_by_",
                  factor_name,
                  ".pdf"),
           plot = plot_name,
           width = 10,
           height = 8)

    plot_name <-
        temp_data %>%
        group_by(across(all_of(factor_name)),
                 cell_type,
                 gene) %>%
        summarize(prop_express = sum(expression > 0) / n(),
                  num_cells = n(),
                  .groups = "drop") %>%
        left_join(cell_type_de) %>%
        mutate(group = paste0(get(factor_name),
                            " (n=",
                            num_cells,
                            ")"),
            p_val_adj = replace_na(p_val_adj, 1),
            sig = ifelse(p_val_adj <= 0.05,
                            "Significant",
                            "Non-significant")) %>%
        ggplot(aes(x = group,
                   y = gene,
                   fill = prop_express)) +
        geom_tile(aes(color = sig),
                  size = 0.5,
                  width = 0.9,
                  height = 0.9) +
        geom_text(aes(label = sprintf("%.1f%%", prop_express * 100)),
                  size = 1,
                  color = "lightgray") +
        facet_wrap(~ cell_type, nrow = 1, scales = "free_x") +
        labs(x = "",
             y = "",
             title = paste0("Percent of Cells Expressing Select Gene")) +
        theme_roberts() +
        scale_fill_continuous(labels = scales::label_percent(accuracy = 2),
                              name = "Percent of\nCells Expressing Gene") +
        scale_color_manual(values = c(plot_cols[1], alpha("black", 0)),
                           name = NULL,
                           breaks = c("Significant", "Non-significant")) +
        theme(legend.position = "right",
              axis.text.x = element_text(angle = 90, hjust = 1)) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(file = paste0("output/figures/growth_factor_perc_expressing_",
                         factor_name,
                         ".pdf"),
        plot = plot_name,
        width = 10,
        height = 8)
}
```

Add primary: S0131


I'd like the following cell types to be in the main figure: dendritic cells, fibroblasts, macrophages, and monocytes And the following genes: AREG, EREG, FGF2, FGF10, FGF18, HBEGF, HGF, IFG1, NRG1, PDGFa, PDGFb, PDGFc, TGFa, TGFb1, and VEGFa
## Make main figures for growth factors
```{r main_figs, dependson='deGrowthFactorMets'}
combined_subset <- qs::qread("output/rdata/combined_subset.qs")

cell_type_de <- read_tsv("output/de/cell_type_de.txt",
                         show_col_types = FALSE)

main_figure_genes <-
    c("Areg", "Ereg", "Fgf2", "Fgf10", "Fgf18", "Hbegf", "Hgf", "Igf1", "Nrg1",
      "Pdgfa", "Pdgfb", "Pdgfc", "Tgfa", "Tgfb1", "Vegfa")

main_figure_cell_types <-
    c("Dendritic cells", "Fibroblasts", "Macrophages", "Monocytes")

main_jitter_cell_types <-
    c("Macrophages", "Monocytes")

combined_subset_main <-
    combined_subset %>%
    subset(cell_type %in% main_figure_cell_types)

combined_subset_main <-
    combined_subset_main[main_figure_genes, ]

for (factor_name in c("group_name", "sample_description")) {
    ############################################################################
    # Plot DE between lung and primary mets by cell type
    temp_data <-
        GetAssayData(combined_subset_main) %>%
        as.matrix() %>%
        t() %>%
        as.data.frame() %>%
        rownames_to_column("cell") %>%
        mutate(cell_type = combined_subset_main$cell_type,
               group_name = combined_subset_main$group_name,
               sample_description = combined_subset_main$sample_description) %>%
        pivot_longer(cols = c(-group_name,
                              -cell_type,
                              -sample_description,
                              -cell),
                     names_to = "gene",
                     values_to = "expression")

    plot_name <-
        temp_data %>%
        filter(cell_type %in% main_jitter_cell_types) %>%
        ggplot(aes(x = gene,
                   y = expression,
                   color = get(factor_name))) +
        geom_point(size = 0.1,
                   alpha = 0.5,
                   position = position_jitterdodge()) +
        facet_wrap(~ cell_type, ncol = 1, scales = "free_y") +
        theme_roberts() +
        labs(x = "",
             y = "Expression",
             title = paste0("Gene Expression Split by ",
                            str_replace(factor_name, "_", " ") %>%
                                str_to_title(),
                            " and Cell Type")) +
        theme(legend.position = "right",
              legend.title = element_blank()) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(paste0("output/figures/jitter_cell_type_by_",
                  factor_name,
                  "_main.pdf"),
           plot = plot_name,
           width = 8,
           height = 4)

    plot_name <-
        temp_data %>%
        group_by(across(all_of(factor_name)),
                 cell_type,
                 gene) %>%
        summarize(prop_express = sum(expression > 0) / n(),
                  num_cells = n(),
                  .groups = "drop") %>%
        left_join(cell_type_de) %>%
        mutate(group = paste0(get(factor_name),
                            " (n=",
                            num_cells,
                            ")"),
            p_val_adj = replace_na(p_val_adj, 1),
            sig = ifelse(p_val_adj <= 0.05,
                            "Significant",
                            "Non-significant")) %>%
        ggplot(aes(x = group,
                   y = gene,
                   fill = prop_express)) +
        geom_tile(aes(color = sig),
                  size = 1,
                  width = 0.9,
                  height = 0.9) +
        geom_text(aes(label = sprintf("%.1f%%", prop_express * 100)),
                  size = 2,
                  color = "lightgray") +
        facet_wrap(~ cell_type, nrow = 1, scales = "free_x") +
        labs(x = "",
             y = "",
             title = paste0("Percent of Cells Expressing Select Gene")) +
        theme_roberts() +
        scale_fill_continuous(labels = scales::label_percent(accuracy = 2),
                              name = "Percent of\nCells Expressing Gene") +
        scale_color_manual(values = c(plot_cols[1], alpha("black", 0)),
                           name = NULL,
                           breaks = c("Significant", "Non-significant")) +
        theme(legend.position = "right",
              axis.text.x = element_text(angle = 90, hjust = 1)) +
        guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

    ggsave(file = paste0("output/figures/growth_factor_perc_expressing_",
                         factor_name,
                         "_main.pdf"),
        plot = plot_name,
        width = 8,
        height = 6)
}
```


```{r, cache=FALSE}
sessionInfo()
```


## UMAP plots
```{r}
#load the sobj
load("output/rdata/sobj_list.rds")

#subset the sobj of interest
umap_subset <-
    sobj_list[c("S0006_mouse",
                "d14_pos_m",
                "d28_pos_m",
                "d49_pos_m",
                "tibia_tumor_h",
                "d14_pos_h",
                "d28_pos_h",
                "d49_pos_h"
                )]


umapplots <- list()
for (object in umap_subset) {
    umapplots[[object$obj_name[[1]]]] <- DimPlot(object) +
            ggtitle(label=object$group) +
            theme_roberts() +
            theme(plot.title = element_text(hjust = 0.5)) +
            NoLegend()
}

# Combine the plots using wrap_plots()
combined_plots <- wrap_plots(umapplots)

# Add a layout specification
combined_plots <- combined_plots + plot_layout(nrow = 2)

# Display the final plot
combined_plots

#save the plot
ggsave("output/figures/combined_plots.pdf",
       width = 4,
       height = 3,
       plot = combined_plots)


#Human only combined umap: this puts them in the same space/comparable
human_only <-
    merge(sobj_list[["tibia_tumor_h"]],
          y = sobj_list[c("d14_pos_h", "d28_pos_h", "d49_pos_h")],
          add.cell.ids = c("Primary tumor", "Early metastatic tumor",
                        "Midpoint metastatic tumor", "Late metastatic tumor"),
          project = "human_only") %>%
          process_seurat()

umap_human_only <- DimPlot(human_only) +
                    ggtitle(label="Primary and metastatic tumor") +
                    theme_roberts() +
                    theme(plot.title = element_text(hjust = 0.5)) +
                    NoLegend()

umap_human_only
```


## GSEA_KEGG

GSEA-KEGG figure showing upregulated genes/pathways as lesions progress from: 
                  S0037 (primary) redo human cells to S0082(early) human cells 
                  S0082 (early) human cell to S0084(midpoint) human cells 
                  S0084(midpoint) human cells to S0093 (late) human cells 
                  S0082(early) human cells to S0093(late) human cells 
```{r}
load("output/rdata/sobj_list.rds")
#we want all of the item in the same space, do combine and run process seurat
human_only <-
    merge(sobj_list[["tibia_tumor_h"]],
          y = sobj_list[c("d14_pos_h",
                          "d28_pos_h",
                          "d49_pos_h")],
          add.cell.ids = c("Primary_tumor",
                           "Early_metastatic_tumor",
                           "Midpoint_metastatic_tumor",
                           "Late_metastatic_tumor"),
          project      =   "human_only") %>%
          process_seurat()

#set the idents
Idents(human_only) <- "group"

#make a table to do the find markers in loop
marker_table <- tribble(~item1, ~item2,
                        "Primary_tumor", "Early_metastatic_tumor",
                        "Early_metastatic_tumor", "Midpoint_metastatic_tumor",
                        "Midpoint_metastatic_tumor", "Late_metastatic_tumor",
                        "Late_metastatic_tumor", "Primary_tumor")


#make a list that you can append to
markers_list <- list()

#FindMarkers since we comaring across only two groups
for (i in 1:nrow(marker_table)) {
        df <- FindMarkers(human_only,
                         ident.1 = marker_table$item1[i],
                         ident.2 = marker_table$item2[i],
                         min.pct = 0.1,
                         logfc.threshold = 0)
        filename <- paste0(marker_table$item1[i],
                            "_",
                            marker_table$item2[i],
                            ".tsv")
        write_tsv(as.data.frame(df), file = filename)
        markers_list[[paste0(marker_table$item1[i],
                            "_",
                            marker_table$item2[i])]] <- df
}


write_tsv(as.as.data.frame(markers_list), "markers_file.tsv")







#FindMarkers since we comaring across only two groups
for (i in 1:nrow(marker_table)) {
    markers_list[[paste0(marker_table$item1[i], "_", marker_table$item2[i])]] <-
                        FindMarkers(human_only,
                         ident.1 = marker_table$item1[i],
                         ident.2 = marker_table$item2[i],
                         min.pct = 0.1,
                         logfc.threshold = 0)

}

write_tsv(as.as.data.frame(markers_list), "markers_file.tsv")


```


